vendor/github.com/openshift/library-go/test/library/etcdverticalscaling/helpers.go

```diff
@@ -0,0 +1,282 @@
+			machinePhase := pointer.StringDeref(machine.Status.Phase, "")
+			if machinePhase != "Running" {
+				return false, fmt.Errorf("%q machine is in unexpected %q state, expected Running", machine.Name, machinePhase)
+			}
+			if !hasMachineDeletionHook(&machine) {
+				return false, fmt.Errorf("%q machine doesn't have required deletion hooks", machine.Name)
+			}
+		}
+		return true, nil
+	})
+}
+
+func recoverClusterToInitialStateIfNeeded(ctx context.Context, t TestingT, machineClient machinev1beta1client.MachineInterface) error {
+	machineList, err := machineClient.List(ctx, metav1.ListOptions{LabelSelector: masterMachineLabelSelector})
+	if err != nil {
+		return err
+	}
+
+	var machineNames []string
+	for _, machine := range machineList.Items {
+		machineNames = append(machineNames, machine.Name)
+	}
+
+	t.Logf("checking if there are any excessive machines in the cluster (created by a previous test), expected cluster size is 3, found %v machines: %v", len(machineList.Items), machineNames)
+	for _, machine := range machineList.Items {
+		if strings.HasSuffix(machine.Name, "-clone") {
+			err := machineClient.Delete(ctx, machine.Name, metav1.DeleteOptions{})
+			if err != nil {
+				return fmt.Errorf("failed removing the machine: %q, err: %v", machine.Name, err)
+			}
+			t.Logf("successfully deleted an excessive machine %q from the API (perhaps, created by a previous test)", machine.Name)
+		}
+	}
+
+	return nil
+}
+
+// EnsureMembersCount simply counts the current etcd members, it doesn't evaluate health conditions or any other attributes (i.e. name) of individual members
+// this method won't fail immediately on errors, this is useful during scaling down operation until the feature can ensure this operation to be graceful
+func EnsureMembersCount(t TestingT, etcdClientFactory EtcdClientCreator, expectedMembersCount int) error {
+	waitPollInterval := 15 * time.Second
+	waitPollTimeout := 10 * time.Minute
+	t.Logf("Waiting up to %s for the cluster to reach the expected member count of %v", waitPollTimeout.String(), expectedMembersCount)
+
+	if err := wait.Poll(waitPollInterval, waitPollTimeout, func() (bool, error) {
+		etcdClient, closeFn, err := etcdClientFactory.NewEtcdClient()
+		if err != nil {
+			t.Logf("failed to get etcd client, will retry, err: %v", err)
+			return false, nil
+		}
+		defer closeFn()
+		defer etcdClient.Close() // ignore the errors
+
+		ctx, cancel := context.WithTimeout(context.TODO(), 15*time.Second)
+		defer cancel()
+		memberList, err := etcdClient.MemberList(ctx)
+		if err != nil {
+			t.Logf("failed to get the member list, will retry, err: %v", err)
+			return false, nil
+		}
+
+		var memberNames []string
+		for _, member := range memberList.Members {
+			memberNames = append(memberNames, member.Name)
+		}
+		if len(memberNames) != expectedMembersCount {
+			t.Logf("unexpected number of etcd members, expected exactly %d, got: %v, current members are: %v", expectedMembersCount, len(memberNames), memberNames)
```

Comment:
Since https://github.com/openshift/cluster-etcd-operator/pull/758 will attempt to promote learners as soon as possible, I'm thinking we should make `EnsureMembersCount` => `EnsureVotingMembersCount()` to have a more strict check for voting members only. Otherwise if you have 3 voting + 1 learner, then `EnsureMembersCount()` still passes.  

That way step 2 above would be more accurate in reporting an error if say there is a failure to promote:
```Go
// step 2: wait until a new member shows up and check if it is healthy
		//         and until all kube-api servers have reached the same revision
		//         this additional step is the best-effort of ensuring they
		//         have observed the new member before disruption
		err = scalingtestinglibrary.EnsureVotingMembersCount(g.GinkgoT(), etcdClientFactory, 4)
		o.Expect(err).ToNot(o.HaveOccurred())
		memberName, err := scalingtestinglibrary.MachineNameToEtcdMemberName(ctx, oc.KubeClient(), machineClient, machineName)
		o.Expect(err).ToNot(o.HaveOccurred())
		err = scalingtestinglibrary.EnsureHealthyMember(g.GinkgoT(), etcdClientFactory, memberName)
		o.Expect(err).ToNot(o.HaveOccurred())
		g.GinkgoT().Log("waiting for api servers to stabilize on the same revision")
		err = testlibraryapi.WaitForAPIServerToStabilizeOnTheSameRevision(g.GinkgoT(), oc.KubeClient().CoreV1().Pods("openshift-kube-apiserver"))
		o.Expect(err).ToNot(o.HaveOccurred())
```

Failing at `EnsureVotingMembersCount()` tells us more accurately that a learner wasn't promoted than trying to figure out why `WaitForAPIServerToStabilizeOnTheSameRevision()` failed later.

Reply:
Also given `EnsureInitialClusterState()` uses `EnsureMembersCount()`:
```Go
// EnsureInitialClusterState makes sure the cluster state is expected, that is, has only 3 running machines and exactly 3 voting members
```
This needs to be voting members only.